_published/0004-core-lib-repo.md

```diff
@@ -0,0 +1,198 @@
+func main() {
+    var version Version
+    config.UnmarshalYAML(versionFile, &version, config.Options{})
+
+    log.Infof("sample-program version=%s", version.AppVersion)
+
+    var config Config
+    configFile, _ := ioutil.ReadFile("config.yaml")
+    config.UnmarshalYAML(configFile, &config, config.Options{
+      AllowEnvironmentVariables: true,
+    })
+
+    log.SetConfig(config.Logging)
+
+    log.Info("Successfully read config.")
+
+    wharfDefFile, _ := ioutil.ReadFile(".wharf-ci.yml")
+    wharfDef, _ := wharfyml.ParseDefinition(wharfDefFile)
+    log.Infof("Read a .wharf-ci.yml file that contains %d stages.", len(wharfDef.Stages))
+
+    r := gin.Default()
+
+    // func ApplyConfig(engine *gin.Engine, ginutils.Config)
+    ginutils.ApplyConfig(r, config.API)
+
+    // func AddVersionEndpoint(engine *gin.Engine, version interface{})
+    ginutils.AddVersionEndpoint(r, version)
+
+    r.Run()
+}
+```
+
+```yaml
+# config.yaml
+logging:
+  level: debug
+  blacklist:
+    - gin
+    - gorm
+api:
+  allowCors: true
+
+db:
+  host: localhost
+  port: 5432
+  username: postgres
+  password: changeit # Can be overritten with WHARF_DB_PASSWORD environment variable
+```
+
+```yaml
+# version.yaml
+$schema: https://github.com/iver-wharf/wharf-core/raw/master/pkg/app/version-schema.json
+appVersion: v1.0.0
+buildGitCommit: 5971d3b585a722536730c39a22aa3148993f2985
+buildRef: 123
+buildDate: 2021-05-12T13:55:00+02:00
+```
+
+## Compatibility
+
+In architectural terms, this utility repository has to be "stable". Meaning it
+will be a hassle to update as there will be so many components relying on it.
+
+This may not be that big of an issue for added features though. It may induce
+issues from time to time where we have to make 4 PRs each time we update the
+logging library, but I [Kalle] don't think that's a major issue as those
+**components should not rely on the logic from the utilities repo to
+cooperate.**
+
+## Alternative solutions
+
+Placing this inside the <https://github.com/iver-wharf/wharf-api> repository.
+While this would work for most of the logic here, it does not play well later
+when the `cmd` project wants to take use of these utilities as well, such as
+the logging.
+
+## Future possibilities
+
+Moving the `.wharf-ci.yml` parsing code out from the `cmd` repo to a utility
+repo will allow the `wharf-api` repo to take use of it and be able to parse
+those definition files inside the API.
+
+It also allows for a unified way of configuring the services. If we can set a
+unified convention of setting configs via files and being able to override them
+with environment variables, then that would lift the operations-experience of
+using Wharf to a much better level. Heavily inspired by the `appsettings.json`
+solution for [configuring ASP.NET Core apps](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-5.0)
+
+## Unresolved questions
+
+1. What should be the name of the repo?
```

Comment:
I am voting for option `wharf-utils`

Reply:
I've added your vote to the RFC itself. You're currently outnumbered, but there's chance for more voters :)